### PR TITLE
Implementation of parallel requests for the block explorer

### PR DESCRIPTION
**Pull request by itsyaasir**
# Issue ticket

Issue ticket link: <>

### Code conventions

- [ ] I have performed a self-review of my code
- [ ] Every new function is documented
- [ ] Object names are auto explicative

### Security

- [ ] The PR does not break APIs backward compatibility
- [ ] The PR does not break the stable storage backward compatibility

### Testing

- [ ] Every function is properly unit tested
- [ ] I have added integration tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] IC endpoints are always tested through the `canister_call!` macro

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    && apt-get clean -y \
+    && rm -rf /var/lib/apt/lists/*
+
+COPY --from=builder /app/target/release/evm-block-extractor-server /app/evm-block-extractor-server
+
+EXPOSE 8080
+
+ENTRYPOINT ["./evm-block-extractor-server"]
+
+CMD ["-d", "testnet", "-s", "0.0.0.0:8080"]
```
Comment (ufoscout) [inline Dockerfile:33]: why `testnet`?

File: Cargo.toml [hunk 2]
```diff
@@ -61,8 +61,12 @@ sha2 = "0.10"
 sha3 = "0.10"
 tempfile = "3"
 thiserror = "1.0"
-tokio = { version = "1.24", features = ["macros", "rt"] }
+tokio = { version = "1.24", features = ["macros", "rt", "signal"] }
 zip = "0.6"
+gcp-bigquery-client = "0.18"
+axum = { version = "0.7", features = ["json", "macros"] }
```
Comment (ufoscout) [inline Cargo.toml:67]: Is `axum`` used somewhere?

File: src/evm-block-extractor/src/storage_clients/gcp_big_query.rs [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+    /// Can be mainnet/testnet
+    dataset_id: String,
+}
+
+impl BigQueryBlockChain {
+    /// Creates a new BigQuery client
+    /// The service account key should be stored in the
+    /// `GCP_BLOCK_EXTRACTOR_SA_KEY` environment variable
+    pub async fn new(dataset_id: String) -> anyhow::Result<Self> {
+        let sa_key = std::env::var("GCP_BLOCK_EXTRACTOR_SA_KEY")
```
Comment (ufoscout) [inline src/evm-block-extractor/src/storage_clients/gcp_big_query.rs:28]: Clap can read default values from env variables too. Please, remove the `GCP_BLOCK_EXTRACTOR_SA_KEY` from here and create a proper entry in the clap configuration file for it.

File: src/evm-block-extractor/src/bin/server.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    module.merge(EthServer::into_rpc(eth))?;
+
+    log::info!("Server started on {}", server.local_addr()?);
+
+    let handle = server.start(module);
+
+    let (stopped_snd, stopped_recv) = oneshot::channel();
+
+    tokio::spawn(graceful_shutdown(stopped_snd));
+    stopped_recv.await?;
```
Comment (GraDKh) [inline src/evm-block-extractor/src/bin/server.rs:42]: Why do you need all this stuff?
Isn't it equivalent to just to
```
match tokio::signal::ctrl_c().await {
        Ok(_) => ,
        Err(err) => log::error!("Failed to listen for shutdown signal: {err}"),
    }
```

File: src/evm-block-extractor/src/bin/server.rs
Comment (ufoscout) [inline src/evm-block-extractor/src/bin/server.rs:1]: Please add a readme file that explains what is this and how to use it. Same for the extractor

**Review by ufoscout — CHANGES_REQUESTED** [review]

File: src/evm-block-extractor/src/block_extractor.rs [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+                log::info!("block result: {:?}", result);
+
+                match result {
+                    Ok(Ok(block)) => {
+                        let mut blockchain = blockchain.lock().await;
+
+                        blockchain.insert_block(&block).await?;
+
+                        let transactions = block.transactions;
+                        for chunk in transactions.chunks(batch_size) {
```
Comment (GraDKh) [inline src/evm-block-extractor/src/block_extractor.rs:81]: 1. Are we sure that the `batch_size` number of receipts will fit into 2 mb limit?
2. Do we need receipts at all? CC @blutooth @ufoscout

File: src/evm-block-extractor/src/block_extractor.rs [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+                }
+
+                drop(permit);
+                Ok(())
+            });
+
+            tasks.push(task);
+        }
+
+        for task in tasks {
```
Comment (GraDKh) [inline src/evm-block-extractor/src/block_extractor.rs:105]: [minor] You could use `join_all`

Comment (itsyaasir) [reply src/evm-block-extractor/src/block_extractor.rs:105]: Yeah, we currently don't use the `futures` crate.

File: Dockerfile (hunk 1 above)
Comment (itsyaasir) [reply Dockerfile:33]: I was trying out the testnet, but this argument should either take testnet or mainnet only

File: Dockerfile
[the remark this answers is not in the document]
Comment (itsyaasir) [reply Dockerfile:1]: I was this docker to publish the block exteactor server images to Google Cloud Run

[the remark this answers is not in the document]
Comment (itsyaasir) [reply Dockerfile:1]: We can't put in the folder you suggested as the `cargo chef` doesn't support building the recipe from a workspace.

File: Cargo.toml (hunk 2 above)
Comment (itsyaasir) [reply Cargo.toml]: Removed

File: Dockerfile
[the remark this answers is not in the document]
Comment (itsyaasir) [reply Dockerfile:1]: Done

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+RUN cargo chef cook --release --recipe-path recipe.json
+
+COPY . .
+RUN cargo build --release --bin evm-block-extractor-server
+
+FROM debian:bookworm-slim AS runtime
+
+WORKDIR /app
+RUN apt-get update -y \
+    && apt-get install -y --no-install-recommends openssl ca-certificates \
```
[the remark this answers is not in the document]
Comment (itsyaasir) [reply Dockerfile]: Removed

File: src/evm-block-extractor/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    && apt-get clean -y \
+    && rm -rf /var/lib/apt/lists/*
+
+COPY --from=builder /app/target/release/evm-block-extractor-server /app/evm-block-extractor-server
+
+EXPOSE 8080
+
+ENTRYPOINT ["./evm-block-extractor-server"]
+
+CMD ["-d", "testnet", "-s", "0.0.0.0:8080"]
```
Comment (ufoscout) [inline src/evm-block-extractor/Dockerfile:33]: The `testnet` is still hardcoded. You should use an `ENV` instead

File: src/evm-block-extractor/src/bin/server.rs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    let handle = server.start(module);
+
+    match tokio::signal::ctrl_c().await {
+        Ok(_) => {
+            log::info!("Received shutdown signal");
+        }
+        Err(err) => log::error!("Failed to listen for shutdown signal: {err}"),
+    }
+
+    handle.stop()?;
```
Comment (ufoscout) [inline src/evm-block-extractor/src/bin/server.rs:59]: Are you sure this is graceful shutdown? the doc says `Tell the server to stop without waiting for the server to stop.`

File: src/evm-block-extractor/src/bin/server.rs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+    log::info!("Server stopped gracefully");
+
+    Ok(())
+}
+
+fn init_logger() -> anyhow::Result<()> {
+    let env = env_logger::Env::default().filter_or(env_logger::DEFAULT_FILTER_ENV, "info");
```
Comment (ufoscout) [inline src/evm-block-extractor/src/bin/server.rs:67]: this is a minor, but we usually have the log settings in the clap config

File: src/evm-block-extractor/src/block_extractor.rs (hunk 5 above)
Comment (ufoscout) [reply src/evm-block-extractor/src/block_extractor.rs:81]: Yes, we decided to store the receipts too

File: src/evm-block-extractor/src/block_extractor.rs (hunk 6 above)
Comment (ufoscout) [reply src/evm-block-extractor/src/block_extractor.rs:105]: If you don't use `join_all`, then the requests are performed sequentially instead of in parallel

File: src/evm-block-extractor/src/storage_clients/gcp_big_query.rs [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+        let client = Client::from_service_account_key(service_account, false).await?;
+
+        Ok(Self {
+            client,
+            project_id: PROJECT_ID.to_string(),
+            dataset_id: dataset_id.to_string(),
+        })
+    }
+
+    async fn execute_query<T: for<'de> serde::Deserialize<'de>>(
```
Comment (ufoscout) [inline src/evm-block-extractor/src/storage_clients/gcp_big_query.rs:39]: How did you test this one? Are you aware of BigQueryEmulator? (`docker pull ghcr.io/goccy/bigquery-emulator:latest`)

File: src/evm-block-extractor/src/storage_clients/gcp_big_query.rs [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+            Err(anyhow::anyhow!("No data found for the query"))
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl BlockChainDB for BigQueryBlockChain {
+    async fn get_block_by_number(&self, block_number: u64) -> anyhow::Result<Block<Transaction>> {
+        let query = format!(
+            "SELECT body FROM `{project_id}.{dataset_id}.{table_id}` WHERE id = {block_number}",
```
Comment (ufoscout) [inline src/evm-block-extractor/src/storage_clients/gcp_big_query.rs:65]: Out of curiosity, is there any protection against SQL injection in this gcp client?

**Review by ufoscout — CHANGES_REQUESTED** [review]

File: src/evm-block-extractor/src/main.rs [hunk 13]
```diff
@@ -52,121 +43,56 @@ async fn main() -> anyhow::Result<()> {
...
+    );
 
-fn init_logger() -> anyhow::Result<()> {
-    env_logger::init();
+    let end_block = extractor.latest_block_number().await?;
+    log::debug!("latest block number: {}", end_block);
 
-    Ok(())
-}
+    let start_block = end_block.saturating_sub(MAX_EVM_BLOCKS);
```
Comment (GraDKh) [inline src/evm-block-extractor/src/main.rs:76]: Can there be not loaded blocks older than `start_block`?

Comment (itsyaasir) [reply src/evm-block-extractor/src/main.rs:76]: Can you rephrase the question please ?

File: src/evm-block-extractor/src/storage_clients/gcp_big_query.rs (hunk 12 above)
Comment (itsyaasir) [reply src/evm-block-extractor/src/storage_clients/gcp_big_query.rs:65]: BigQuery parameterized queries are  immune to SQL injection by [design](https://cloud.google.com/bigquery/docs/parameterized-queries).

Comment (ufoscout) [reply src/evm-block-extractor/src/storage_clients/gcp_big_query.rs:65]: Well... that was my point, these queries are not parameterized :)

File: src/evm-block-extractor/src/bin/server.rs (hunk 9 above)
Comment (itsyaasir) [reply src/evm-block-extractor/src/bin/server.rs]: Changed

File: src/evm-block-extractor/src/storage_clients/gcp_big_query.rs (hunk 12 above)
Comment (itsyaasir) [reply src/evm-block-extractor/src/storage_clients/gcp_big_query.rs:65]: But this is client, when the request is sent over to gcp as a query parameter, it is immune to SQL injection.

Comment (ufoscout) [reply src/evm-block-extractor/src/storage_clients/gcp_big_query.rs:65]: You are not using query parameters because all query data is hardcoded in the query string. You should use placeholders to avoid SQL injection (see: https://github.com/lquerel/gcp-bigquery-client/blob/bd82fd16cdac389507975743478840103747e6db/examples/pagination.rs#L13 )

File: src/evm-block-extractor/src/block_extractor.rs (hunk 5 above)
Comment (ufoscout) [reply src/evm-block-extractor/src/block_extractor.rs:81]: The 2mb limit will be addressed in a dedicated PR

File: src/evm-block-extractor/src/main.rs [hunk 14]
```diff
@@ -52,121 +51,57 @@ async fn main() -> anyhow::Result<()> {
...
```
Comment (ufoscout) [inline src/evm-block-extractor/src/main.rs:85]: The start block should be calculated based on the last block we already stored

File: src/evm-block-extractor/src/storage_clients/gcp_big_query.rs [hunk 15]
```diff
@@ -0,0 +1,393 @@
...
+    pub async fn init(&self) -> anyhow::Result<()> {
+        let dataset = Dataset::new(&self.project_id, &self.dataset_id);
+
+        // Define tables with their respective schemas
+        let tables = [
+            (
+                BLOCKS_TABLE_ID,
+                vec![
+                    TableFieldSchema::integer("id"),
+                    TableFieldSchema::string("body"),
```
Comment (ufoscout) [inline src/evm-block-extractor/src/storage_clients/gcp_big_query.rs:74]: BigQuery has a specific JSON column type, we should use it

File: src/evm-block-extractor/src/storage_clients/gcp_big_query.rs [hunk 16]
```diff
@@ -0,0 +1,393 @@
...
+        let mut insert_request = TableDataInsertAllRequest::new();
+
+        let block_id = block
+            .number
+            .ok_or(anyhow::anyhow!("Block number not found"))?
+            .as_u64();
+
+        let block_row = BlockRow {
+            id: block_id,
+            body: serde_json::to_string(&block)?,
```
Comment (ufoscout) [inline src/evm-block-extractor/src/storage_clients/gcp_big_query.rs:230]: We should double checking the docs, I think we should use a JSON here (and everywhere else) instead of a string

**Review by ufoscout — CHANGES_REQUESTED** [review]

File: src/evm-block-extractor/tests/big_query_client_it.rs [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+    let mut blockchain = Box::new(
+        BigQueryBlockChain::new_with_client(
+            project_id.clone(),
+            dataset_id.clone(),
+            gcp_client.clone(),
+        )
+        .unwrap(),
+    );
+
+    blockchain.init().await.unwrap();
```
Comment (ufoscout) [inline src/evm-block-extractor/tests/big_query_client_it.rs:184]: Add a test where the init is executed in a row to be sure it does not fail if the tables exist

File: src/evm-block-extractor/src/storage_clients/gcp_big_query.rs (hunk 15 above)
Comment (itsyaasir) [reply src/evm-block-extractor/src/storage_clients/gcp_big_query.rs:74]: Changed

File: src/evm-block-extractor/src/storage_clients/gcp_big_query.rs (hunk 16 above)
Comment (itsyaasir) [reply src/evm-block-extractor/src/storage_clients/gcp_big_query.rs:230]: In this line we are converting the body into json formatted string.

**Review by ufoscout — APPROVED** [review]

**Review by veeso — APPROVED** [review]

**Review by F3kilo — APPROVED** [review]

**Review by GraDKh — APPROVED** [review]
